onnxruntime/contrib_ops/cuda/bert/attention_impl.cu

```diff
@@ -94,14 +94,28 @@ __device__ inline void Softmax(const int sequence_length, const int valid_length
   }
 
   const auto sum = BlockReduce(tmp_storage).Reduce(thread_data_sum, cub::Sum());
+  bool is_sum_zero = false;
+
   if (threadIdx.x == 0) {
-    sum_reverse_block = 1.f / sum;
+    if (sum != 0) {
+      sum_reverse_block = 1.f / sum;
+    } else {
+      is_sum_zero = true;
+    }
   }
   __syncthreads();
 
   for (int i = threadIdx.x; i < sequence_length; i += TPB) {
     const int index = offset + i;
-    const float val = (i < num_valid) ? expf(float(input[index]) - max_block) * sum_reverse_block : 0.f;
+    float val = 0.f;
+    if (i < num_valid) {
+      if (!is_sum_zero) {
+        val = expf(float(input[index]) - max_block) * sum_reverse_block;      
+      } else {
+        val = (1.f) / sequence_length;
+      }   
```

Comment:
naive question: why 1/sequence_length?

Reply:
I ll update this PR. I think rather than "handling" sum == 0, we need to find out why sum is becoming zero. sum should not be zero (if implemented right) because while computing softmax over a bunch of values - x1,...,xn - we first subtract the max - meaning one of them becomes 0 and exp(0) == 1 and exp of the rest has to be >= 0 (because exp does not give negative values) - so the sum of 1 and bunch of non-negative values should not be zero in the first place. Understanding why this happened will probably solve this entire issue

Reply:
No need to set 1/sequence_length to hide the problem.

I suggest we just set sum_reverse_block to 0.0 in line 103 to avoid NaN when sum is zero (Thus all values become zero after softmax).

We can do similarly in cpu code.

Reply:
I have made the change more simple and added comments to justify - please take a look.